src/Adapter/Customer/QueryHandler/GetCustomerInformationHandler.php

```diff
@@ -0,0 +1,590 @@
+            (bool) $customer->optin
+        );
+
+        return new PersonalInformation(
+            $customer->firstname,
+            $customer->lastname,
+            $customer->email,
+            $customer->isGuest(),
+            $socialTitle,
+            $birthday,
+            $registrationDate,
+            $lastUpdateDate,
+            $lastVisitDate,
+            $this->getCustomerRankBySales($customer->id),
+            $customerShop->name,
+            $customerLanguage->name,
+            $customerSubscriptions,
+            (bool) $customer->active
+        );
+    }
+
+    /**
+     * @param int $customerId
+     *
+     * @return int|null customer rank or null if customer is not ranked
+     */
+    private function getCustomerRankBySales($customerId)
+    {
+        $sql = 'SELECT SUM(total_paid_real) FROM ' . _DB_PREFIX_ . 'orders WHERE id_customer = ' . (int) $customerId . ' AND valid = 1';
+
+        if ($totalPaid = Db::getInstance()->getValue($sql)) {
+            $sql = '
+                SELECT SQL_CALC_FOUND_ROWS COUNT(*)
+                FROM ' . _DB_PREFIX_ . 'orders
+                WHERE valid = 1
+                    AND id_customer != ' . (int) $customerId . '
+                GROUP BY id_customer
+                HAVING SUM(total_paid_real) > ' . (int) $totalPaid
+            ;
+
+            Db::getInstance()->getValue($sql);
+
+            return (int) Db::getInstance()->getValue('SELECT FOUND_ROWS()') + 1;
+        }
+
+        return null;
+    }
+
+    /**
+     * @param Customer $customer
+     *
+     * @return OrdersInformation
+     */
+    private function getCustomerOrders(Customer $customer)
+    {
+        $validOrders = [];
+        $invalidOrders = [];
+
+        $orders = Order::getCustomerOrders($customer->id, true);
+        $totalSpent = 0;
+
+        foreach ($orders as $order) {
+            $order['total_paid_real_not_formated'] = $order['total_paid_real'];
+            $order['total_paid_real'] = Tools::displayPrice(
+                $order['total_paid_real'],
+                new Currency((int) $order['id_currency'])
+            );
+
+            if (!isset($order['order_state'])) {
+                $order['order_state'] = $this->translator->trans(
+                    'There is no status defined for this order.',
+                    [],
+                    'Admin.Orderscustomers.Notification'
+                );
+            }
+
+            $customerOrderInformation = new OrderInformation(
+                (int) $order['id_order'],
+                Tools::displayDate($order['date_add']),
+                $order['payment'],
+                $order['order_state'],
+                (int) $order['nb_products'],
+                $order['total_paid_real']
+            );
+
+            if ($order['valid']) {
+                $validOrders[] = $customerOrderInformation;
+                $totalSpent += $order['total_paid_real_not_formated'] / $order['conversion_rate'];
```

Comment:
Do not assume `$order['conversion_rate']` will be an integer different from zero. Assert it 😉

Reply:
it must be. :) it comes directly from db and it's `decimal(13,6)`.

Reply:
Still cant assume it. Do not rely on your database structure. Trust no one 😛

Reply:
well, it was working for years (probably :D), so i think it safe to leave it like it is, at some point it will be totally refactored along with other adapters. :)

however, if you really want me to assert it, then what fallback action should it take if it fails?

Reply:
We can assume that if the data is not valid, although it comes from the database, it must be
- either a developer modifying the code, so it's a mistake
- or a database structure not valid
In both cases we should throw an Exception as these are critical scenarii so we can fail the request. [Crash early](https://www.quora.com/What-does-crash-early-mean-in-software-engineering).

Reply:
ok, but what exception should be thrown here? its part of domain, but i dont think its neither `CustomerException` nor `DomainException` would be correct here, maybe global `LogicException`, wdyt?